File: urbanairship-core/src/main/java/com/urbanairship/analytics/data/EventDao.java
package com.urbanairship.analytics.data;

import com.urbanairship.Logger;
import com.urbanairship.util.UAStringUtil;

import java.util.List;

import androidx.annotation.RestrictTo;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;

/**
 * Event data access object.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
@Dao
public abstract class EventDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void insert(EventEntity event);

    @Transaction
    @Query("SELECT * FROM events ORDER BY id ASC")
    public abstract List<EventEntity> get();

    @Transaction
    @Query("SELECT id, eventId, data FROM events ORDER BY id ASC LIMIT :limit")
    public abstract List<EventEntity.EventIdAndData> getBatch(int limit);

    @Transaction
    public void deleteBatch(List<EventEntity.EventIdAndData> events) {
        for (EventEntity.EventIdAndData event : events) {
            delete(event.eventId);
        }
    }

    @Query("DELETE FROM events WHERE eventId = :eventId")
    abstract void delete(String eventId);

    @Delete()
    public abstract void delete(EventEntity... events);

    @Query("DELETE FROM events")
    public abstract void deleteAll();

    @Query("SELECT COUNT(*) FROM events")
    public abstract int count();

    @Query("SELECT SUM(eventSize) FROM events")
    public abstract int databaseSize();

    @Query("SELECT sessionId FROM events ORDER BY id ASC LIMIT 1")
    abstract String oldestSessionId();

    @Query("DELETE FROM events WHERE sessionId = :sessionId")
    abstract int deleteSession(String sessionId);

    @Transaction
    public void trimDatabase(int maxDatabaseSize) {
        while (databaseSize() > maxDatabaseSize) {
            String sessionId = oldestSessionId();
            if (UAStringUtil.isEmpty(sessionId)) {
                return;
            }

            Logger.debug("Event database size exceeded. Deleting oldest session: %s", sessionId);

            int deleted = deleteSession(sessionId);
            Logger.debug("Deleted %d rows with session ID %s", deleted, sessionId);

            if (deleted == 0) {
                return;
            }
        }
    }
}

File: urbanairship-core/src/main/java/com/urbanairship/analytics/data/EventDao.kt
package com.urbanairship.analytics.data

import androidx.annotation.RestrictTo
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import com.urbanairship.UALog
import com.urbanairship.analytics.data.EventEntity.EventIdAndData

/**
 * Event data access object.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
@Dao
internal abstract class EventDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    internal abstract suspend fun insert(event: EventEntity)

    @Transaction
    @Query("SELECT * FROM events ORDER BY id ASC")
    internal abstract suspend fun get(): List<EventEntity>

    @Transaction
    @Query("SELECT id, eventId, data FROM events ORDER BY id ASC LIMIT :limit")
    internal abstract suspend fun getBatch(limit: Int): List<EventIdAndData>

    @Transaction
    internal open suspend fun deleteBatch(events: List<EventIdAndData>) {
        for (event in events) {
            delete(event.eventId)
        }
    }

    @Query("DELETE FROM events WHERE eventId = :eventId")
    internal abstract suspend fun delete(eventId: String)

    @Delete
    internal abstract suspend fun delete(vararg events: EventEntity)

    @Query("DELETE FROM events")
    internal abstract suspend fun deleteAll()

    @Query("SELECT COUNT(*) FROM events")
    internal abstract suspend fun count(): Int

    @Query("SELECT SUM(eventSize) FROM events")
    internal abstract suspend fun databaseSize(): Int

    @Query("SELECT sessionId FROM events ORDER BY id ASC LIMIT 1")
    internal abstract suspend fun oldestSessionId(): String?

    @Query("DELETE FROM events WHERE sessionId = :sessionId")
    internal abstract suspend fun deleteSession(sessionId: String): Int

    @Transaction
    internal open suspend fun trimDatabase(maxDatabaseSize: Int) {
        while (databaseSize() > maxDatabaseSize) {
            val sessionId = oldestSessionId() ?: return

            if (sessionId.isEmpty()) {
                return
            }

            UALog.d("Event database size exceeded. Deleting oldest session: $sessionId", )

            val deleted = deleteSession(sessionId)
            UALog.d("Deleted $deleted rows with session ID $sessionId")

            if (deleted == 0) {
                return
            }
        }
    }
}
